File: modules/analysis/templates/04-qual/qual_analysis.R
# Anonymize, segment, and scaffold qualitative coding from real text files.

`%||%` <- function(a, b) if (is.null(a) || length(a) == 0) b else a
find_shared <- function() {
  cmd <- commandArgs(FALSE); file_arg <- sub("^--file=", "", cmd[grep("^--file=", cmd)][1] %||% "")
  p <- if (nzchar(file_arg)) dirname(normalizePath(file_arg, mustWork = FALSE)) else getwd()
  repeat { candidate <- file.path(p, "_shared", "r_common.R"); if (file.exists(candidate)) return(candidate); np <- dirname(p); if (identical(np, p)) stop("Cannot locate templates/_shared/r_common.R"); p <- np }
}
source(find_shared())
args <- parse_common_args(); ensure_dirs(args$out_root)
text_dir <- if (!is.na(args$data)) args$data else "."
files <- if (dir.exists(text_dir)) list.files(text_dir, pattern = "\\.(txt|md)$", full.names = TRUE) else if (file.exists(text_dir)) text_dir else character()
if (!length(files)) {
  report <- write_md(file.path(args$out_root, "qual/memos", paste0("qual-blockers-", Sys.Date(), ".md")), "Qual Blockers", c("阻断" = "缺少可读取的 .txt/.md 文本材料。"))
  log_run(args, "04-qual", "blocked", report, "未发现文本材料。")
  quit(status = 2)
}
rows <- list()
for (f in files) {
  txt <- paste(readLines(f, warn = FALSE, encoding = "UTF-8"), collapse = "\n")
  txt <- gsub("[[:alnum:]._%+-]+@[[:alnum:].-]+\\.[A-Za-z]{2,}", "[EMAIL]", txt)
  txt <- gsub("\\b1[3-9][0-9]{9}\\b", "[PHONE]", txt)
  anon <- file.path(args$out_root, "qual/anonymized", paste0(tools::file_path_sans_ext(basename(f)), "-anonymized.txt"))
  writeLines(txt, anon, useBytes = TRUE)
  parts <- unlist(strsplit(txt, "\\n\\s*\\n"))
  parts <- parts[nzchar(trimws(parts))]
  for (i in seq_along(parts)) rows[[length(rows) + 1]] <- data.frame(source_id = basename(f), excerpt_id = i, anonymized_excerpt = substr(parts[[i]], 1, 1000), code = "", memo = "")
}
coded <- write_csv_safe(do.call(rbind, rows), file.path(args$out_root, "qual/coded-data", paste0("coded-excerpts-", Sys.Date(), ".csv")))
codebook <- write_csv_safe(data.frame(code = "", definition = "", inclusion = "", exclusion = "", example = ""), file.path(args$out_root, "qual/codebooks", paste0("codebook-", Sys.Date(), ".csv")))
log_run(args, "04-qual", "ok", c(coded, codebook), "已真实去标识化和分段，不伪造主题结论。")

File: modules/analysis/templates/04-qual/qual_analysis.py
#!/usr/bin/env python3
"""Anonymize, segment, and scaffold qualitative coding from real text files."""

from __future__ import annotations

import argparse
import re
import sys
from datetime import date
from pathlib import Path


def _load_common():
    for parent in Path(__file__).resolve().parents:
        if (parent / "_shared").exists():
            sys.path.insert(0, str(parent / "_shared"))
            import python_common  # type: ignore

            return python_common
    raise RuntimeError("Cannot locate templates/_shared/python_common.py")


common = _load_common()


def anonymize(text: str) -> str:
    text = re.sub(r"[\w.+-]+@[\w-]+\.[\w.-]+", "[EMAIL]", text)
    text = re.sub(r"\b1[3-9]\d{9}\b", "[PHONE]", text)
    text = re.sub(r"\b\d{15,18}[0-9Xx]\b", "[ID]", text)
    return text


def main() -> int:
    parser = common.add_common_args(argparse.ArgumentParser(description=__doc__))
    parser.add_argument("--text-dir", default=None, help="Directory of .txt/.md files.")
    args = parser.parse_args()
    p = common.paths(args.out_root)
    text_dir = Path(args.text_dir or args.data or ".")
    files = [x for pat in ("*.txt", "*.md") for x in text_dir.glob(pat)] if text_dir.is_dir() else ([text_dir] if text_dir.exists() else [])
    if not files:
        report = common.write_markdown(p["qual"] / "memos" / f"qual-blockers-{date.today().isoformat()}.md", "Qual Blockers", {"阻断": "缺少可读取的 .txt/.md 文本材料。"})
        common.log_run(args, "04-qual", "blocked", [report], "未发现文本材料。")
        return 2
    coded_rows = []
    for f in files:
        text = anonymize(f.read_text(encoding="utf-8", errors="ignore"))
        anon_path = p["qual"] / "anonymized" / f"{f.stem}-anonymized.txt"
        anon_path.write_text(text, encoding="utf-8")
        parts = [x.strip() for x in re.split(r"\n\s*\n", text) if x.strip()]
        for i, part in enumerate(parts, 1):
            coded_rows.append({"source_id": f.stem, "excerpt_id": i, "anonymized_excerpt": part[:1000], "code": "", "memo": ""})
    codebook = common.write_csv(p["qual"] / "codebooks" / f"codebook-{date.today().isoformat()}.csv", [{"code": "", "definition": "", "inclusion": "", "exclusion": "", "example": ""}])
    coded = common.write_csv(p["qual"] / "coded-data" / f"coded-excerpts-{date.today().isoformat()}.csv", coded_rows)
    reliability = common.write_markdown(p["qual"] / "reliability" / f"reliability-{date.today().isoformat()}.md", "Qualitative Reliability", {"复核": "已完成去标识化和分段；主题、信度和结论需由人工编码或复核后填写。"})
    common.log_run(args, "04-qual", "ok", [codebook, coded, reliability], "已真实去标识化和分段，不伪造主题结论。")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
